azblk.c

```diff
@@ -0,0 +1,1006 @@
+		}
+	}
+
+	switch (action) {
+	case CURL_POLL_IN:
+		uv_poll_start(&context->poll_handle, UV_READABLE,
+			      azblk_curl_perform);
+		break;
+	case CURL_POLL_OUT:
+		uv_poll_start(&context->poll_handle, UV_WRITABLE,
+			      azblk_curl_perform);
+		break;
+	case CURL_POLL_INOUT:
+		uv_poll_start(&context->poll_handle, UV_READABLE | UV_WRITABLE,
+			      azblk_curl_perform);
+		break;
+	case CURL_POLL_REMOVE:
+		if (socketp) {
+			context = (struct azblk_socket_context *)socketp;
+
+			uv_poll_stop(&context->poll_handle);
+			azblk_destroy_socket_context(context);
+			curl_multi_assign(ddev->curl_multi, s, NULL);
+		}
+		break;
+	}
+
+	return 0;
+}
+
+void azblk_dev_loop(void *arg)
+{
+	struct azblk_dev *ddev = (struct azblk_dev *)arg;
+	int ret;
+
+	ret = uv_run(&ddev->loop, UV_RUN_DEFAULT);
+
+	uv_walk(&ddev->loop, azblk_loop_cleanup, NULL);
+
+	uv_run(&ddev->loop, UV_RUN_DEFAULT);
+
+	ret = uv_loop_close(&ddev->loop);
+	if (ret == UV_EBUSY)
+		tcmu_dev_warn(ddev->dev, "Not all libuv handles are closed\n");
+}
+
+static bool azblk_parse_config(struct azblk_dev *ddev)
+{
+	char *cfgstring;
+	char *tcm_dev_name;
+	config_setting_t *setting;
+	config_t cfg;
+	char *str;
+	char *err_msg;
+	int i, count;
+	bool found_name = false;
+	bool found_url = false;
+
+	cfgstring = tcmu_dev_get_cfgstring(ddev->dev);
+	if (strncmp(cfgstring, "azblk/", 6) != 0) {
+		tcmu_dev_err(ddev->dev,
+			     "Invalid cfgstring format: %s\n",
+			     cfgstring);
+		return false;
+	}
+
+	str = (char *)cfgstring + 6;
+
+	if (strncmp(str, "none", 4) == 0)
+		strcpy(ddev->cfg.config_file, "/etc/tcmu/azblk.conf");
+	else
+		strcpy(ddev->cfg.config_file, str);
+
+	tcmu_info("Reading config file %s\n", ddev->cfg.config_file);
+
+	config_init(&cfg);
+
+	if (!config_read_file(&cfg, ddev->cfg.config_file)) {
+		asprintf(&err_msg, "File: %s:%d - %s\n",
+			 ddev->cfg.config_file, config_error_line(&cfg),
+			 config_error_text(&cfg));
+		goto error;
+	}
+
+	setting = config_lookup(&cfg, "device");
+	if (!setting) {
+		asprintf(&err_msg, "File %s is malformed\n",
+			 ddev->cfg.config_file);
+		goto error;
+	}
+
+	tcm_dev_name = tcmu_dev_get_tcm_dev_name(ddev->dev);
+
+	count = config_setting_length(setting);
+
+	for (i = 0; i < count; ++i) {
+		const char *name, *sas, *lease, *url;
+		int read_only, len;
+		config_setting_t *device = config_setting_get_elem(setting, i);
+
+		if (!(config_setting_lookup_string(device, "name", &name))) {
+			asprintf(&err_msg,
+				 "File %s: device name required\n",
+				 ddev->cfg.config_file);
+			goto error;
+		}
+
+		if (strcmp(name, tcm_dev_name) != 0)
+			continue;
+
+		found_name = true;
+
+		if (config_setting_lookup_string(device, "sas", &sas)) {
+			len = strlen(sas);
+			if (len > (AZ_SAS_LEN - 1)) {
+				asprintf(&err_msg,
+					 "File %s: device %s sas must be less than %d characters\n",
+					 ddev->cfg.config_file,
+					 tcm_dev_name,
+					 AZ_SAS_LEN);
+				goto error;
+			}
+			strncpy(ddev->cfg.sas, sas, len);
```

Comment:
Here we'd better to use the strlcpy as Mike has introduce it from BSD locally.
https://github.com/open-iscsi/tcmu-runner/blob/master/strlcpy.c